qiskit/transpiler/passes/utils/gate_direction.py

```diff
@@ -103,46 +106,97 @@ def run(self, dag):
-
+        trivial_layout = Layout.generate_trivial_layout(*dag.qregs.values())
+        layout_map = trivial_layout.get_virtual_bits()
         if len(dag.qregs) > 1:
             raise TranspilerError(
                 "GateDirection expects a single qreg input DAG,"
                 "but input DAG had qregs: {}.".format(dag.qregs)
             )
+        if self.target is None:
+            cmap_edges = set(self.coupling_map.get_edges())
+            if not cmap_edges:
+                return dag
 
-        trivial_layout = Layout.generate_trivial_layout(*dag.qregs.values())
-        layout_map = trivial_layout.get_virtual_bits()
-        dist_matrix = self.coupling_map.distance_matrix
+            self.coupling_map.compute_distance_matrix()
 
-        for node in dag.two_qubit_ops():
-            control = node.qargs[0]
-            target = node.qargs[1]
+            dist_matrix = self.coupling_map.distance_matrix
 
-            physical_q0 = layout_map[control]
-            physical_q1 = layout_map[target]
+            for node in dag.two_qubit_ops():
+                control = node.qargs[0]
+                target = node.qargs[1]
 
-            if dist_matrix[physical_q0, physical_q1] != 1:
-                raise TranspilerError(
-                    "The circuit requires a connection between physical "
-                    "qubits %s and %s" % (physical_q0, physical_q1)
-                )
+                physical_q0 = layout_map[control]
+                physical_q1 = layout_map[target]
+
+                if dist_matrix[physical_q0, physical_q1] != 1:
+                    raise TranspilerError(
+                        "The circuit requires a connection between physical "
+                        "qubits %s and %s" % (physical_q0, physical_q1)
+                    )
+
+                if (physical_q0, physical_q1) not in cmap_edges:
+                    if node.name == "cx":
+                        dag.substitute_node_with_dag(node, self._cx_dag)
+                    elif node.name == "ecr":
+                        dag.substitute_node_with_dag(node, self._ecr_dag)
+                    elif node.name == "rzx":
+                        dag.substitute_node_with_dag(node, self._rzx_dag(*node.op.params))
+                    else:
+                        raise TranspilerError(
+                            f"Flipping of gate direction is only supported "
+                            f"for CX, ECR, and RZX at this time, not {node.name}."
+                        )
+        else:
+            # TODO: Work with the gate instances and only use names as look up keys.
+            # This will require iterating over the target names to build a mapping
+            # of names to gates that implement CXGate, ECRGate, RZXGate (including
+            # fixed angle variants)
+            cx_weak_set = None
+            if "cx" in self.target:
+                cx_weak_set = {frozenset(qarg) for qarg in self.target["cx"]}
```

Comment:
That said, since these are only two qubit gates, perhaps this could be simplified by something like:
```python
if (q0, q1) in target[op_name]:
    continue
elif (q1, q0) in target[op_name]:
    dag.substitute_node_with_dag(reverse_op)
else:
    raise TranspilerError(...)
```

Reply:
That is much simpler, I've been stuck thinking in an overlapping set mindset, I'll refactor this now

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7227/commits/fe21d8d5eb407c67976b46c924323b7d0637e5b6